diff -r 4987f8a59d8a linux/drivers/media/dvb/Kconfig
--- a/linux/drivers/media/dvb/Kconfig	Sat Jun 18 00:53:46 2011 +0300
+++ b/linux/drivers/media/dvb/Kconfig	Sun Jun 26 07:30:18 2011 +0200
@@ -84,6 +84,10 @@
 	depends on DVB_CORE && PCI && I2C
 	source "drivers/media/dvb/ngene/Kconfig"
 
+comment "Supported HDHOMERUN"
+	depends on DVB_CORE && PCI && I2C
+	source "drivers/media/dvb/hdhomerun/Kconfig"
+
 comment "Supported DVB Frontends"
 	depends on DVB_CORE
 source "drivers/media/dvb/frontends/Kconfig"
diff -r 4987f8a59d8a linux/drivers/media/dvb/Makefile
--- a/linux/drivers/media/dvb/Makefile	Sat Jun 18 00:53:46 2011 +0300
+++ b/linux/drivers/media/dvb/Makefile	Sun Jun 26 07:30:18 2011 +0200
@@ -16,6 +16,7 @@
 		pt1/		\
 		mantis/		\
 		smi2032/	\
+		hdhomerun/	\
 		ngene/
 
 obj-$(CONFIG_DVB_FIREDTV)	+= firewire/
diff -r 4987f8a59d8a linux/drivers/media/dvb/frontends/Kconfig
--- a/linux/drivers/media/dvb/frontends/Kconfig	Sat Jun 18 00:53:46 2011 +0300
+++ b/linux/drivers/media/dvb/frontends/Kconfig	Sun Jun 26 07:30:18 2011 +0200
@@ -625,6 +625,11 @@
 config DVB_DUMMY_FE
 	tristate "Dummy frontend driver"
 	default n
+
+config DVB_HDHOMERUN
+	tristate "Hdhomerun FE"
+	default n
+
 endmenu
 
 endif
diff -r 4987f8a59d8a linux/drivers/media/dvb/frontends/Makefile
--- a/linux/drivers/media/dvb/frontends/Makefile	Sat Jun 18 00:53:46 2011 +0300
+++ b/linux/drivers/media/dvb/frontends/Makefile	Sun Jun 26 07:30:18 2011 +0200
@@ -4,6 +4,7 @@
 
 EXTRA_CFLAGS += -Idrivers/media/dvb/dvb-core/
 EXTRA_CFLAGS += -Idrivers/media/common/tuners/
+EXTRA_CFLAGS += -DS2_LIPLIANIN
 
 s921-objs := s921_module.o s921_core.o
 stb0899-objs = stb0899_drv.o stb0899_algo.o
@@ -85,3 +86,4 @@
 obj-$(CONFIG_DVB_EC100) += ec100.o
 obj-$(CONFIG_DVB_DS3000) += ds3000.o
 obj-$(CONFIG_DVB_MB86A16) += mb86a16.o
+obj-$(CONFIG_DVB_HDHOMERUN) += dvb_hdhomerun_fe.o
